File: tasking/sources/taskgraph.5.c
/*
* @@name:	taskgraph.5
* @@type:	C
* @@operation:	run
* @@expect:	success
* @@version:	omp_6.0
*/
#include <stdio.h>

#define N 100
#define M 8
#define NG 4

int xi[NG][N];
int yi[NG][N];
int zi[NG][N];
int sums[M]={0};

void get_vals(int b[], const int n)
{
   for (int j = 0; j < n; j++) {
      b[j] = j;
   }
}

void do_compute(int z[], int x[], int y[], const int n)
{
   for (int j = 0; j < n; j++) {
      z[j] = x[j] * y[j] - x[j];
   }
}

void do_parallel_work(const int i)
{
   const int input_index = i % NG;
   int *x = xi[input_index];
   int *y = yi[input_index];
   int *z = zi[input_index];
   #pragma omp parallel masked
      #pragma omp taskgraph graph_id(input_index)
      {
	 // Task A
	 #pragma omp task firstprivate(saved: x) depend(out: *x)
	 get_vals(x, N);
	 // Task B
	 #pragma omp task firstprivate(saved: y) depend(out: *y)
	 get_vals(y, N);
	 // Task C
	 #pragma omp task firstprivate(saved: x,y,z) depend(in: *x,*y)
	 {
	   do_compute(z, x, y, N);
	   #pragma simd reduction(+:sums[i])
	   for (int j = 0; j < N; j++)
	     sums[i] += z[j];
	 }
      }
}

int check_sum(const int n)
{
   return  ( (n-1) * n * (2*n-4) ) / 6;
}

int main()
{
   for (int i = 0; i < M; i++) {
      do_parallel_work(i);
   }
   for (int i = 0; i < M; i++) {
      if (sums[i] != check_sum(N)) {
         printf("check failed\n");
         return 1;
      }
   }
   printf("check passed\n");
   return 0;
}

File: tasking/sources/taskgraph.5.f90
! @@name:	taskgraph.5
! @@type:	F-free
! @@operation:	run
! @@expect:	success
! @@version:	omp_6.0

program taskgraph_5
   integer, parameter :: N = 100
   integer, parameter :: M = 8
   integer, parameter :: NG = 4
   integer :: sums(M) = (/ (0, i=1,M) /)
   integer, target :: xi(N,NG), yi(N,NG), zi(N,NG)
   integer :: i, j

   do i = 1, M
      call do_parallel_work(i)
   end do

   do i = 1, M
      if (sums(i) /= check_sum(n)) then
         print *, "check failed"
         stop 1
      end if
   end do
   print *, "check passed"

   contains

   subroutine do_parallel_work(i)
      integer, intent(in) :: i
      integer, pointer :: x(:), y(:), z(:)
      integer :: input_index
      input_index = mod(i-1, NG) + 1
      x => xi(:,input_index)
      y => yi(:,input_index)
      z => zi(:,input_index)
      !$omp parallel masked
         !$omp taskgraph graph_id(input_index)
            ! Task A
            !$omp task firstprivate(saved: x) depend(out: x)
               call get_vals(x, N)
            !$omp end task
            ! Task B
            !$omp task firstprivate(saved: y) depend(out: y)
               call get_vals(y, N)
            !$omp end task
            ! Task C
            !$omp task firstprivate(saved: x,y,z) depend(in: x,y)
               call do_compute(z, x, y, N)
               !$omp simd reduction(+:sums(i))
               do j = 1, N
                  sums(i) = sums(i) + z(j)
               end do
            !$omp end task
         !$omp end taskgraph
      !$omp end parallel masked
   end subroutine

   subroutine get_vals(b, n)
      integer, intent(out) :: b(*)
      integer, intent(in) :: n
      integer :: j
      do j = 1, n
         b(j) = j
      end do
   end subroutine

   subroutine do_compute(z, x, y, n)
      integer, intent(out) :: z(*)
      integer, intent(in) :: x(*), y(*), n
      integer :: j
      do j = 1, n
         z(j) = x(j) * y(j) - x(j)
      end do
   end subroutine

   function check_sum(n) result(cs)
      integer, intent(in) :: n
      integer :: cs
      cs = ( n * (n+1) * (2*n-2) ) / 6
   end function
end program

